File: C++/Sort_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *sortList(ListNode *head) {
        ListNode *ptr = head;
        int len = 0;
        while (ptr != NULL) {
            len ++;
            ptr = ptr->next;
        }
        if (len == 0) return NULL;
        return solve(head, len);
    }
    ListNode *solve(ListNode *head, int len){
        if (len == 1) return head;
        ListNode *ptr = head;
        int count = 1;
        while (count < len/2){
            ptr = ptr->next;
            count ++;
        }
        ListNode *h2 = solve(ptr->next, len - count);
        ptr->next = NULL;
        ListNode *h1 = solve(head, count);
        return merge(h1, h2);
    }
    ListNode *merge(ListNode *h1, ListNode *h2){
        ListNode *head = NULL;
        ListNode *ptr1 = h1, *ptr2 = h2;
        ListNode *ptr = head, *cand = NULL;
        while (ptr1 != NULL && ptr2 != NULL) {
            if (ptr1->val < ptr2->val) {
                cand = ptr1;
                ptr1 = ptr1->next;
            }
            else {
                cand = ptr2;
                ptr2 = ptr2->next;
            }
            if (head == NULL) head = cand;
            else ptr->next = cand;
            ptr = cand;
        }
        if (ptr1 != NULL) ptr->next = ptr1;
        if (ptr2 != NULL) ptr->next = ptr2;
        return head;
    }
};

File: Java/Sort_List.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        int n = 0;
        ListNode ptr = head;
        while (ptr != null) {
            n ++;
            ptr = ptr.next;
        }
        return sortListRecursively(head, n);
    }
    
    private ListNode sortListRecursively(ListNode head, int n){
        if (n <= 1) return head;
        ListNode ptr = head;
        int len = 1;
        while (len < n/2) {
            ptr = ptr.next;
            len ++;
        }
        ListNode secondHalf = ptr.next;
        ptr.next = null;
        return mergeLists(sortListRecursively(head, n/2), sortListRecursively(secondHalf, n-n/2));
    }
    
    private ListNode mergeLists(ListNode h1, ListNode h2){
        ListNode ret = null, retPtr = null, ptr1 = h1, ptr2 = h2;
        while (ptr1 != null && ptr2 != null) {
            if (ptr1.val < ptr2.val) {
                if (retPtr != null) retPtr.next = ptr1;
                retPtr = ptr1;
                ptr1 = ptr1.next;
            }
            else {
                if (retPtr != null) retPtr.next = ptr2;
                retPtr = ptr2;
                ptr2 = ptr2.next;
            }
            if (ret == null) ret = retPtr;
        }
        if (ptr1 != null) retPtr.next = ptr1;
        if (ptr2 != null) retPtr.next = ptr2;
        return ret;
    }
}
